File: src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.java
/*
 * Copyright 2015 Austin Keener, Michael Ritter, Florian Spieß, and the JDA contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.dv8tion.jda.api.interactions.commands;

import net.dv8tion.jda.internal.utils.EntityString;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Objects;
import java.util.StringJoiner;

/**
 * Represents a slash command mention, such as {@code </ban soft:1021082477038678126>}
 */
public class SlashCommandReference implements ICommandReference
{
    private final long id;
    private final String name;
    private final String subcommand;
    private final String subcommandGroup;

    public SlashCommandReference(@Nonnull String name, @Nullable String subcommandGroup, @Nullable String subcommand, long id)
    {
        this.name = name;
        this.subcommandGroup = subcommandGroup;
        this.subcommand = subcommand;
        this.id = id;
    }

    @Override
    public long getIdLong()
    {
        return id;
    }

    @Nonnull
    @Override
    public String getName()
    {
        return name;
    }

    /**
     * Returns the subcommand of the slash command
     *
     * @return the subcommand of the slash command
     */
    @Nullable
    public String getSubcommandName()
    {
        return subcommand;
    }

    /**
     * Returns the subcommand group of the slash command
     *
     * @return the subcommand group of the slash command
     */
    @Nullable
    public String getSubcommandGroup()
    {
        return subcommandGroup;
    }

    @Nonnull
    @Override
    public String getFullCommandName()
    {
        final StringJoiner joiner = new StringJoiner(" ");
        joiner.add(name);
        if (subcommandGroup != null)
            joiner.add(subcommandGroup);
        if (subcommand != null)
            joiner.add(subcommand);

        return joiner.toString();
    }

    @Override
    public String toString()
    {
        return new EntityString(this)
                .setName(getFullCommandName())
                .toString();
    }

    @Override
    public boolean equals(Object o)
    {
        if (this == o) return true;
        if (!(o instanceof SlashCommandReference)) return false;

        SlashCommandReference that = (SlashCommandReference) o;

        return id == that.id
            && name.equals(that.name)
            && Objects.equals(subcommand, that.subcommand)
            && Objects.equals(subcommandGroup, that.subcommandGroup);
    }

    @Override
    public int hashCode()
    {
        return Objects.hash(id, name, subcommand, subcommandGroup);
    }
}

File: src/main/java/net/dv8tion/jda/api/interactions/commands/SlashCommandReference.kt
/*
 * Copyright 2015 Austin Keener, Michael Ritter, Florian Spieß, and the JDA contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dv8tion.jda.api.interactions.commands

import net.dv8tion.jda.internal.utils.EntityString
import java.util.*
import javax.annotation.Nonnull

/**
 * Represents a slash command mention, such as `</ban soft:1021082477038678126>`
 */
class SlashCommandReference(
    @get:Nonnull
    @param:Nonnull override val name: String,
    /**
     * Returns the subcommand group of the slash command
     *
     * @return the subcommand group of the slash command
     */
    val subcommandGroup: String?,
    /**
     * Returns the subcommand of the slash command
     *
     * @return the subcommand of the slash command
     */
    val subcommandName: String?, override val idLong: Long
) : ICommandReference {

    @get:Nonnull
    override val fullCommandName: String
        get() {
            val joiner = StringJoiner(" ")
            joiner.add(name)
            if (subcommandGroup != null) joiner.add(subcommandGroup)
            if (subcommandName != null) joiner.add(subcommandName)
            return joiner.toString()
        }

    override fun toString(): String {
        return EntityString(this)
            .setName(fullCommandName)
            .toString()
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is SlashCommandReference) return false
        val that = o
        return idLong == that.idLong && name == that.name && subcommandName == that.subcommandName && subcommandGroup == that.subcommandGroup
    }

    override fun hashCode(): Int {
        return Objects.hash(idLong, name, subcommandName, subcommandGroup)
    }
}
